sklearn/model_selection/validate.py

```diff
@@ -0,0 +1,905 @@
+                                                      fit_params)
+                            for train, test in cv.split(X, y, labels))
+    p = np.concatenate([p for p, _ in preds_blocks])
+    locs = np.concatenate([loc for _, loc in preds_blocks])
+    if not _check_is_permutation(locs, _num_samples(X)):
+        raise ValueError('cross_val_predict only works for partitions')
+    preds = p.copy()
+    preds[locs] = p
+    return preds
+
+
+def _fit_and_predict(estimator, X, y, train, test, verbose, fit_params):
+    """Fit estimator and predict values for a given dataset split.
+
+    Read more in the :ref:`User Guide <validate>`.
+
+    Parameters
+    ----------
+    estimator : estimator object implementing 'fit' and 'predict'
+        The object to use to fit the data.
+
+    X : array-like of shape at least 2D
+        The data to fit.
+
+    y : array-like, optional, default: None
+        The target variable to try to predict in the case of
+        supervised learning.
+
+    train : array-like, shape (n_train_samples,)
+        Indices of training samples.
+
+    test : array-like, shape (n_test_samples,)
+        Indices of test samples.
+
+    verbose : integer
+        The verbosity level.
+
+    fit_params : dict or None
+        Parameters that will be passed to ``estimator.fit``.
+
+    Returns
+    -------
+    preds : sequence
+        Result of calling 'estimator.predict'
+
+    test : array-like
+        This is the value of the test parameter
+    """
+    # Adjust length of sample weights
+    fit_params = fit_params if fit_params is not None else {}
+    fit_params = dict([(k, _index_param_value(X, v, train))
+                      for k, v in fit_params.items()])
+
+    X_train, y_train = _safe_split(estimator, X, y, train)
+    X_test, _ = _safe_split(estimator, X, y, test, train)
+
+    if y_train is None:
+        estimator.fit(X_train, **fit_params)
+    else:
+        estimator.fit(X_train, y_train, **fit_params)
+    preds = estimator.predict(X_test)
+    return preds, test
+
+
+def _check_is_permutation(locs, n):
+    """Check whether locs is a reordering of the array np.arange(n)
+
+    Parameters
+    ----------
+    locs : ndarray
+        integer array to test
+    n : int
+        number of expected elements
+
+    Returns
+    -------
+    is_partition : bool
+        True iff sorted(locs) is range(n)
+    """
+    if len(locs) != n:
+        return False
+    hit = np.zeros(n, bool)
+    hit[locs] = True
+    if not np.all(hit):
+        return False
+    return True
+
+
+def _index_param_value(X, v, indices):
+    """Private helper function for parameter value indexing."""
+    if not _is_arraylike(v) or _num_samples(v) != _num_samples(X):
+        # pass through: skip indexing
+        return v
+    if sp.issparse(v):
+        v = v.tocsr()
+    return safe_indexing(v, indices)
+
+
+def permutation_test_score(estimator, X, y, labels=None, cv=None,
```

Comment:
pl ignore the earlier comment on a related discussion... (I didn't see this before)...

and okay!